pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +246,66 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
-
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	MAC_Address macAddr = {0};
+#if defined(AF_LINK)
+	struct ifaddrs *adapterInfo, *po;
+	int error = getifaddrs(&adapterInfo);
+	if (error)
+	{
+		Console.Error("DEV9: Failed to get adapter information %s", error);
+		return std::nullopt;
+	}
+	for (po = adapterInfo; po != NULL; po = po->ifa_next)
```

Comment:
Instead of calling `freeifaddrs()` on every exit path, you can use a scoped guard to release it when the function returns (might need to include `common/ScopedGuard.h`).
i.e.
```cpp
const ScopedGuard adapterInfoGuard = [&adapterInfo]() {
  freeifaddrs(adapterInfo);
});

Reply:
Nice trick!